sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rule/BeamIOPushDownRule.java

```diff
@@ -0,0 +1,236 @@
+
+    relBuilder.filter(newFilter);
+    relBuilder.project(newProjects, calc.getRowType().getFieldNames()); // Preserve named projects
+
+    RelNode result = relBuilder.build();
+
+    call.transformTo(result);
+  }
+
+  /**
+   * Given a {@code RexNode}, find all {@code RexInputRef}s a node or it's children nodes use.
+   *
+   * @param inputRowType {@code RelDataType} used for looking up names of {@code RexInputRef}.
+   * @param startNode A node to start at.
+   * @param usedFields Names of {@code RexInputRef}s are added to this list.
+   */
+  @VisibleForTesting
+  void findUtilizedInputRefs(RelDataType inputRowType, RexNode startNode, Set<String> usedFields) {
+    Queue<RexNode> prerequisites = new ArrayDeque<>();
+    prerequisites.add(startNode);
+
+    // Assuming there are no cyclic nodes, traverse dependency tree until all RexInputRefs are found
+    while (!prerequisites.isEmpty()) {
+      RexNode node = prerequisites.poll();
+
+      if (node instanceof RexCall) { // Composite expression, example: "=($t11, $t12)"
+        RexCall compositeNode = (RexCall) node;
+
+        // Expression from example above contains 2 operands: $t11, $t12
+        prerequisites.addAll(compositeNode.getOperands());
+      } else if (node instanceof RexInputRef) { // Input reference
+        // Find a field in an inputRowType for the input reference
+        int inputFieldIndex = ((RexInputRef) node).getIndex();
+        RelDataTypeField field = inputRowType.getFieldList().get(inputFieldIndex);
+
+        // If we have not seen it before - add it to the list (hash set)
+        usedFields.add(field.getName());
+      }
+    }
+  }
+
+  /**
+   * Recursively reconstruct a {@code RexNode}, mapping old RexInputRefs to new.
+   *
+   * @param node {@code RexNode} to reconstruct.
+   * @param inputRefMapping Mapping from old {@code RexInputRefNode} indexes to new, where list
+   *     index is the new {@code RexInputRefNode} and the value is old {@code RexInputRefNode}.
+   * @return reconstructed {@code RexNode} with {@code RexInputRefNode} remapped to new values.
+   */
+  @VisibleForTesting
+  RexNode reMapRexNodeToNewInputs(RexNode node, List<Integer> inputRefMapping) {
+    if (node instanceof RexInputRef) {
+      int oldInputIndex = ((RexInputRef) node).getIndex();
+      int newInputIndex = inputRefMapping.indexOf(oldInputIndex);
+
+      // Create a new input reference pointing to a new input field
+      return new RexInputRef(newInputIndex, node.getType());
+    } else if (node instanceof RexCall) { // Composite expression, example: "=($t11, $t12)"
+      RexCall compositeNode = (RexCall) node;
+      List<RexNode> newOperands = new ArrayList<>();
+
+      for (RexNode operand : compositeNode.getOperands()) {
+        newOperands.add(reMapRexNodeToNewInputs(operand, inputRefMapping));
+      }
+
+      return compositeNode.clone(compositeNode.getType(), newOperands);
+    }
+
+    // If node is anything else - return it as is (ex: Literal)
```

Comment:
Can you add an assert to back this? Don't want to silently do the wrong thing if Calcite adds a new type.

Reply:
Added a `Preconditions.checkArgument(node instanceof RexLiteral)`.